domain/application/resource/store.go

```diff
@@ -4,64 +4,96 @@
 package resource
 
 import (
-	"context"
-	"fmt"
-	"io"
+    "context"
+    "fmt"
+    "io"
+    "path"
 
-	"github.com/juju/juju/core/objectstore"
-	"github.com/juju/juju/internal/charm/resource"
+    "github.com/juju/juju/core/application"
+    "github.com/juju/juju/core/objectstore"
+    "github.com/juju/juju/internal/charm/resource"
 )
 
 // ResourceStore provides a list of methods necessary for interacting with
 // a store for the resource.
 type ResourceStore interface {
-	// Get returns an io.ReadCloser for data at path, namespaced to the
-	// model.
-	Get(context.Context, string) (io.ReadCloser, int64, error)
+    // Get returns an io.ReadCloser for a resource in the resource store.
+    Get(ctx context.Context, applicationUUID application.ID, name string) (r io.ReadCloser, size int64, err error)
+
+    // Put stores data from io.Reader in the resource store at the
+    // path specified in the resource.
+    Put(ctx context.Context, r io.Reader, resource Resource) error
+
+    // Remove removes a resource from storage.
+    Remove(ctx context.Context, applicationUUID application.ID, name string) error
 }
 
 // FileResourceStore is a ResourceStore for file resource types.
 type FileResourceStore struct {
-	objectStore objectstore.ObjectStore
+    objectStore objectstore.ObjectStore
+}
+
+// Get the specified resource from the object store.
+func (f FileResourceStore) Get(
+    ctx context.Context,
+    applicationUUID application.ID,
+    name string,
+) (io.ReadCloser, int64, error) {
+    path := f.storagePath(name, applicationUUID)
+    return f.objectStore.Get(ctx, path)
+}
+
+// Put the given resource in the object store.
+func (f FileResourceStore) Put(ctx context.Context, r io.Reader, resource Resource) error {
+    path := f.storagePath(resource.Name, resource.ApplicationID)
+    return f.objectStore.PutAndCheckHash(ctx, path, r, resource.Size, resource.Fingerprint.String())
```

Comment:
**question:** I have seen in the DDL that it is possible that their is no fingerprint (resource.hash_type_id == 0 ie `none`:

https://github.com/juju/juju/blob/4464a08035b0ef5844063275a15ef73ed1deae0f/domain/schema/model/sql/0006-objectstore-metadata.sql#L9-L11

Maybe we should have some logic here to call `f.objectStore.Put` in this case ? @hmlanigan : what do you think?

Reply:
I might be wrong, but I think there is only going to be no fingerprint for OCI images, and we should require it for files. In the 3.6 state `Put` was never used directly, only `PutAndCheckHash`.